### PR TITLE
Add full typing

### PR DESCRIPTION
**Pull request by mfbehrens99**

### CODE REVIEW

**Comment by lmas** [comment]
Thanks, been wanting this for a while 🤟 I'm busy until sometime next week so it's going to take a while for me to review.

-- 
Alex

File: opensimplex/opensimplex.py [hunk 1]
```diff
@@ -74,28 +74,28 @@ def _init(seed=DEFAULT_SEED):
 
 
 @njit(cache=True)
-def _extrapolate2(perm, xsb, ysb, dx, dy):
+def _extrapolate2(perm: np.ndarray, xsb: int, ysb: int, dx: float, dy: float) -> np.ndarray:
```
Comment (lmas) [inline opensimplex/opensimplex.py:77]: Why are you mixing ints and floats here? Most everywhere else it's been floats

File: opensimplex/opensimplex.py [hunk 2]
```diff
@@ -74,28 +74,28 @@ def _init(seed=DEFAULT_SEED):
...
     index = perm[(perm[xsb & 0xFF] + ysb) & 0xFF] & 0x0E
     g1, g2 = GRADIENTS2[index : index + 2]
     return g1 * dx + g2 * dy
 
 
 @njit(cache=True)
-def _extrapolate3(perm, perm_grad_index3, xsb, ysb, zsb, dx, dy, dz):
+def _extrapolate3(perm: np.ndarray, perm_grad_index3: np.ndarray, xsb: int, ysb: int, zsb: int, dx: float, dy: float, dz: float) -> np.ndarray:
```
Comment (lmas) [inline opensimplex/opensimplex.py:84]: ints and floats

File: opensimplex/opensimplex.py [hunk 3]
```diff
@@ -74,28 +74,28 @@ def _init(seed=DEFAULT_SEED):
...
     index = perm_grad_index3[(perm[(perm[xsb & 0xFF] + ysb) & 0xFF] + zsb) & 0xFF]
     g1, g2, g3 = GRADIENTS3[index : index + 3]
     return g1 * dx + g2 * dy + g3 * dz
 
 
 @njit(cache=True)
-def _extrapolate4(perm, xsb, ysb, zsb, wsb, dx, dy, dz, dw):
+def _extrapolate4(perm: np.ndarray, xsb: int, ysb: int, zsb: int, wsb: int, dx: float, dy: float, dz: float, dw: float) -> np.ndarray:
```
Comment (lmas) [inline opensimplex/opensimplex.py:91]: ints and floats

File: opensimplex/opensimplex.py (hunk 1 above)
Comment (lmas) [reply opensimplex/opensimplex.py:77]: Nvm, I compared it with the original C source functions and the inputs matched correctly :+1:

File: opensimplex/opensimplex.py [hunk 4]
```diff
@@ -19,40 +19,40 @@ def wrapper(func):
...
+def overflow(x: int) -> int:
     # Since normal python ints and longs can be quite humongous we have to use
     # self hack to make them be able to overflow.
     # Using a np.int64 won't work either, as it will still complain with:
     # "OverflowError: int too big to convert"
     return c_int64(x).value
 
 
-def _init(seed=DEFAULT_SEED):
+def _init(seed: int = DEFAULT_SEED) -> tuple[np.ndarray, np.ndarray]:
```
Comment (lmas) [inline opensimplex/opensimplex.py:55]: This line fails the tests:

```
=====================================================================
ERROR: Failure: TypeError ('type' object is not subscriptable)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/home/lmas/code/public/opensimplex/devenv/lib/python3.8/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/usr/home/lmas/code/public/opensimplex/devenv/lib/python3.8/site-packages/nose/loader.py", line 417, in loadTestsFromName
    module = self.importer.importFromPath(
  File "/usr/home/lmas/code/public/opensimplex/devenv/lib/python3.8/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/home/lmas/code/public/opensimplex/devenv/lib/python3.8/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/usr/local/lib/python3.8/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/usr/local/lib/python3.8/imp.py", line 171, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 702, in _load
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/home/lmas/code/public/opensimplex/tests/test_opensimplex.py", line 11, in <module>
    import opensimplex as simplex
  File "/usr/home/lmas/code/public/opensimplex/opensimplex/__init__.py", line 5, in <module>
    from .opensimplex import OpenSimplex, np
  File "/usr/home/lmas/code/public/opensimplex/opensimplex/opensimplex.py", line 56, in <module>
    def _init(seed: int = DEFAULT_SEED) -> tuple[np.ndarray, np.ndarray]:
TypeError: 'type' object is not subscriptable
```

Comment (lmas) [reply opensimplex/opensimplex.py:55]: Hmm seems to be a v3.9 thing according to https://docs.python.org/3/library/typing.html#typing.Tuple

Comment (lmas) [reply opensimplex/opensimplex.py:55]: Been pondering this and I think maybe it's better not to add type hints to the internal functions (in the file `opensimplex.py`)? It's only going to encourage people to use them, when they're not supposed to.
The type hints for the public API in `__init__.py` is still wanted though!

Also don't have to fix this bug with hacky version imports or something (as I want to keep the version compatibility).

Toughs?

Comment (mfbehrens99) [reply opensimplex/opensimplex.py]: I think type hinting is not encouraging people to use the functions. After all, what is the problem of people using the internal functions. Moreover, it is better for ourselves to have the code type hinted for finding potential bugs.

**Comment by lmas** [comment]
Eh I finally decided to not bother with any quirks and simply removed the type hint for that func for now :shrug:

**Comment by lmas** [comment]
After more consideration and some research, I'm still not convinced of any usefulness of adding type hints to the internals and still think it's just adding unnecessary complexity/line noise with little to no gain at this time.

On the other hand, I gladly accept type hints for the public API as they can serve as a usage hints and hopefully make it clearer how to use the API.

**Comment by lmas** [comment]
Thank you for work!

**Pull request merged**